cmd: add drt-run command

itsbilal opened the pull request:
This command adds drt-run, a relatively simple go-based binary that concurrently executes some workloads as defined in a config YAML file, as well as multiple operation workers that schedule and run operations concurrently. The actual running of workloads and operations is done by subprocesses of `workload` and `roachtest` managed by this command, mostly to reduce code inter-dependencies between the roachtest package and this package.

This command also supports an http endpoint that shows recent log entries from each operation and workload.

Epic: none

Release note: None

Commit:
cmd: add drt-run commmand

This command adds drt-run, a relatively simple go-based binary
that concurrently executes some workloads as defined in a config
YAML file, as well as multiple operation workers that schedule and
run operations concurrently. The actual running of workloads
and operations is done by subprocesses of `workload` and
`roachtest` managed by this command, mostly to reduce code
inter-dependencies between the roachtest package and this package.

This command also supports an http endpoint that shows recent log
entries from each operation and workload.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124093)
<!-- Reviewable:end -->

itsbilal commented:
I think what's here is ready for review, however I'm not sure if I like that the entire package has no tests. What'd be the best way to test this while mocking out any cloud interactions; maybe writing an integration-y test against a `local` roachprod cluster? I'll give that a try.

itsbilal commented:
Also open to splitting out `http.go` into its own PR, if this PR is getting too large. But the good thing is that all the code changes are in one (new) package

herkolategan commented on pkg/cmd/drt-run/operations.go:
```diff
@@ -0,0 +1,314 @@
+		opSpec := &opSpecsSet[rng.Intn(len(opSpecsSet))]
+		r.mu.Lock()
+		lastRun := r.mu.lastRun[opSpec.Name]
+		eligibleForNextRun := lastRun.Add(r.config.Operations.Sets[setIdx].Cadence)
```
Following the logic here I think we could end up in scenarios where an operation of the same kind starts and overlaps with itself. It also brings up another question as to if we should have two exclusions options: One for not running concurrently, and another for supporting concurrency, but not with itself. The latter I think is problematic if an operation is waiting on clean-up (like disk stall or node kill), this logic could start that same operation again, and we'd be in a bad state, I think?

itsbilal replied:
I've made this change and updated the CanRunConcurrently type. PTAL!

itsbilal commented:
@herkolategan I've addressed most of your review points so this should be ready for another look. I might make some other small quality of life improvements such as easier Ctrl+C handling, and I'll take another peek at templating, but otherwise this works!

herkolategan reviewed (approved):
Thanks for making the changes, looks good to me! Although I think it would be good to have a member, or two, from the DRP team also give this a pass.

vidit-bhat commented on pkg/cmd/drt-run/operations.go:
```diff
@@ -0,0 +1,348 @@
+	signal.Notify(sig, os.Interrupt)
+	go func() {
+		<-sig
+		// Signal runners to stop.
```
I think we need to make this a `select` block

```
select {
case <-sig:
case <-ctx.Done():
	return
}
```

See https://github.com/cockroachdb/cockroach/pull/129781

itsbilal replied:
Done.

vidit-bhat commented:
LGTM save one comment. Thanks for all the work!

itsbilal commented:
Thanks for all the reviews!
